File: packages/apidash_mcp/src/oauth/store.ts
/**
 * OAuth 2.1 In-Memory Store with Disk Persistence for Clients
 *
 * Clients are persisted to .oauth-clients.json so registered client_ids
 * survive server restarts (VS Code stores its client_id permanently).
 *
 * Auth codes and tokens remain in-memory only (short-lived by design).
 */

import crypto from "crypto";
import fs     from "fs";
import path   from "path";
import { fileURLToPath } from "url";

// Path to persistence file (same dir as this module → src/oauth/)
const __dirname    = path.dirname(fileURLToPath(import.meta.url));
const CLIENTS_FILE = path.join(__dirname, "..", "..", ".oauth-clients.json");

// ─────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────

export interface OAuthClient {
  client_id: string;
  client_secret?: string;
  redirect_uris: string[];
  client_name?: string;
  grant_types: string[];
  response_types: string[];
  token_endpoint_auth_method: string;
  scope?: string;
  created_at: number;
}

export interface AuthCode {
  code: string;
  client_id: string;
  redirect_uri: string;
  code_challenge: string;
  code_challenge_method: "S256" | "plain";
  scope: string;
  expires_at: number;
}

export interface AccessToken {
  token: string;
  client_id: string;
  scope: string;
  expires_at: number;
  refresh_token?: string;
}

export interface RefreshToken {
  token: string;
  client_id: string;
  scope: string;
}

// ─────────────────────────────────────────────────────────────
// In-memory stores
// ─────────────────────────────────────────────────────────────

// Load persisted clients on startup
function loadClientsFromDisk(): Map<string, OAuthClient> {
  try {
    if (fs.existsSync(CLIENTS_FILE)) {
      const raw  = fs.readFileSync(CLIENTS_FILE, "utf8");
      const arr  = JSON.parse(raw) as OAuthClient[];
      return new Map(arr.map(c => [c.client_id, c]));
    }
  } catch (e) {
    console.error("[oauth] Failed to load clients from disk:", e);
  }
  return new Map();
}

function saveClientsToDisk(map: Map<string, OAuthClient>): void {
  try {
    fs.writeFileSync(CLIENTS_FILE, JSON.stringify([...map.values()], null, 2));
  } catch (e) {
    console.error("[oauth] Failed to persist clients to disk:", e);
  }
}

const clients      = loadClientsFromDisk();
const authCodes    = new Map<string, AuthCode>();
const accessTokens = new Map<string, AccessToken>();
const refreshTokens = new Map<string, RefreshToken>();

// ─────────────────────────────────────────────────────────────
// Client management (RFC 7591)
// ─────────────────────────────────────────────────────────────

export function registerClient(data: Partial<OAuthClient>): OAuthClient {
  const client_id     = crypto.randomUUID();
  const isPublic      = data.token_endpoint_auth_method === "none";
  const client_secret = isPublic ? undefined : crypto.randomBytes(32).toString("hex");

  const client: OAuthClient = {
    client_id,
    client_secret,
    redirect_uris:              data.redirect_uris              ?? [],
    client_name:                data.client_name,
    grant_types:                data.grant_types                ?? ["authorization_code", "refresh_token"],
    response_types:             data.response_types             ?? ["code"],
    token_endpoint_auth_method: data.token_endpoint_auth_method ?? "none",
    scope:                      data.scope                      ?? "mcp",
    created_at: Date.now(),
  };

  clients.set(client_id, client);
  saveClientsToDisk(clients);     // ← persist immediately
  console.log(`[oauth] Registered client: ${client_id} (${client.client_name ?? "unnamed"}) — saved to disk`);
  return client;
}

export function getClient(client_id: string): OAuthClient | undefined {
  return clients.get(client_id);
}

// ─────────────────────────────────────────────────────────────
// Authorization codes (PKCE, 60 s TTL)
// ─────────────────────────────────────────────────────────────

export function createAuthCode(
  data: Omit<AuthCode, "code" | "expires_at">
): AuthCode {
  const code: AuthCode = {
    ...data,
    code:       crypto.randomBytes(32).toString("base64url"),
    expires_at: Date.now() + 60_000,
  };
  authCodes.set(code.code, code);
  return code;
}

/** Atomically consume (read + delete) an auth code. Returns null if missing/expired. */
export function consumeAuthCode(code: string): AuthCode | null {
  const ac = authCodes.get(code);
  if (!ac) return null;
  authCodes.delete(code);
  if (Date.now() > ac.expires_at) return null;
  return ac;
}

// ─────────────────────────────────────────────────────────────
// Access tokens (1 hr TTL) + refresh tokens (rolling)
// ─────────────────────────────────────────────────────────────

export function createAccessToken(
  data: Omit<AccessToken, "token" | "expires_at">
): AccessToken {
  const token        = crypto.randomBytes(32).toString("base64url");
  const refreshToken = crypto.randomBytes(32).toString("base64url");

  const at: AccessToken = {
    ...data,
    token,
    expires_at:    Date.now() + 3_600_000, // 1 hr
    refresh_token: refreshToken,
  };

  accessTokens.set(token, at);
  refreshTokens.set(refreshToken, {
    token:     refreshToken,
    client_id: data.client_id,
    scope:     data.scope,
  });

  return at;
}

export function validateAccessToken(token: string): AccessToken | null {
  const at = accessTokens.get(token);
  if (!at) return null;
  if (Date.now() > at.expires_at) {
    accessTokens.delete(token);
    return null;
  }
  return at;
}

/** One-time-use: consume refresh token and return its data. */
export function consumeRefreshToken(token: string): RefreshToken | null {
  const rt = refreshTokens.get(token);
  if (!rt) return null;
  refreshTokens.delete(token);
  return rt;
}

export function revokeToken(token: string): boolean {
  if (accessTokens.delete(token))  return true;
  if (refreshTokens.delete(token)) return true;
  return false;
}

File: packages/apidash_mcp/lib/src/oauth/store.dart
import 'dart:convert';
import 'dart:io';
import 'package:crypto/crypto.dart';
import 'dart:math';

// Generate a random string like base64url without padding
String _generateRandomString(int length) {
  final rand = Random.secure();
  final values = List<int>.generate(length, (i) => rand.nextInt(256));
  return base64UrlEncode(values).replaceAll('=', '');
}

final _clientsFile = File('${Platform.environment['HOME']}/.local/share/apidash/.oauth-clients.json');

class OAuthClient {
  final String clientId;
  final String? clientSecret;
  final List<String> redirectUris;
  final String? clientName;
  final List<String> grantTypes;
  final List<String> responseTypes;
  final String tokenEndpointAuthMethod;
  final String scope;
  final int createdAt;

  OAuthClient({
    required this.clientId,
    this.clientSecret,
    required this.redirectUris,
    this.clientName,
    required this.grantTypes,
    required this.responseTypes,
    required this.tokenEndpointAuthMethod,
    required this.scope,
    required this.createdAt,
  });

  Map<String, dynamic> toJson() => {
    'client_id': clientId,
    'client_secret': clientSecret,
    'redirect_uris': redirectUris,
    'client_name': clientName,
    'grant_types': grantTypes,
    'response_types': responseTypes,
    'token_endpoint_auth_method': tokenEndpointAuthMethod,
    'scope': scope,
    'created_at': createdAt,
  };

  factory OAuthClient.fromJson(Map<String, dynamic> json) => OAuthClient(
    clientId: json['client_id'] as String,
    clientSecret: json['client_secret'] as String?,
    redirectUris: (json['redirect_uris'] as List).cast<String>(),
    clientName: json['client_name'] as String?,
    grantTypes: (json['grant_types'] as List).cast<String>(),
    responseTypes: (json['response_types'] as List).cast<String>(),
    tokenEndpointAuthMethod: json['token_endpoint_auth_method'] as String,
    scope: json['scope'] as String? ?? 'mcp',
    createdAt: json['created_at'] as int,
  );
}

class AuthCode {
  final String code;
  final String clientId;
  final String redirectUri;
  final String codeChallenge;
  final String codeChallengeMethod;
  final String scope;
  final int expiresAt;

  AuthCode({
    required this.code,
    required this.clientId,
    required this.redirectUri,
    required this.codeChallenge,
    required this.codeChallengeMethod,
    required this.scope,
    required this.expiresAt,
  });
}

class AccessToken {
  final String token;
  final String clientId;
  final String scope;
  final int expiresAt;
  final String? refreshToken;

  AccessToken({
    required this.token,
    required this.clientId,
    required this.scope,
    required this.expiresAt,
    this.refreshToken,
  });
}

class RefreshToken {
  final String token;
  final String clientId;
  final String scope;

  RefreshToken({
    required this.token,
    required this.clientId,
    required this.scope,
  });
}

Map<String, OAuthClient> _loadClientsFromDisk() {
  try {
    if (_clientsFile.existsSync()) {
      final raw = _clientsFile.readAsStringSync();
      final arr = jsonDecode(raw) as List;
      final map = <String, OAuthClient>{};
      for (final item in arr) {
        final c = OAuthClient.fromJson(item as Map<String, dynamic>);
        map[c.clientId] = c;
      }
      return map;
    }
  } catch (e) {
    print('[oauth] Failed to load clients from disk: $e');
  }
  return {};
}

void _saveClientsToDisk(Map<String, OAuthClient> map) {
  try {
    _clientsFile.writeAsStringSync(jsonEncode(map.values.map((c) => c.toJson()).toList()),
        mode: FileMode.writeOnly);
  } catch (e) {
    print('[oauth] Failed to persist clients to disk: $e');
  }
}

final _clients = _loadClientsFromDisk();
final _authCodes = <String, AuthCode>{};
final _accessTokens = <String, AccessToken>{};
final _refreshTokens = <String, RefreshToken>{};

OAuthClient registerClient(Map<String, dynamic> data) {
  final clientId = _generateRandomString(16);
  final isPublic = data['token_endpoint_auth_method'] == 'none';
  final clientSecret = isPublic ? null : _generateRandomString(32);

  final client = OAuthClient(
    clientId: clientId,
    clientSecret: clientSecret,
    redirectUris: (data['redirect_uris'] as List?)?.cast<String>() ?? [],
    clientName: data['client_name'] as String?,
    grantTypes: (data['grant_types'] as List?)?.cast<String>() ?? ['authorization_code', 'refresh_token'],
    responseTypes: (data['response_types'] as List?)?.cast<String>() ?? ['code'],
    tokenEndpointAuthMethod: (data['token_endpoint_auth_method'] as String?) ?? 'none',
    scope: data['scope'] as String? ?? 'mcp',
    createdAt: DateTime.now().millisecondsSinceEpoch,
  );

  _clients[clientId] = client;
  _saveClientsToDisk(_clients);
  print('[oauth] Registered client: $clientId (${client.clientName ?? "unnamed"}) — saved to disk');
  return client;
}

OAuthClient? getClient(String clientId) {
  return _clients[clientId];
}

AuthCode createAuthCode({
  required String clientId,
  required String redirectUri,
  required String codeChallenge,
  required String codeChallengeMethod,
  required String scope,
}) {
  final code = _generateRandomString(32);
  final ac = AuthCode(
    code: code,
    clientId: clientId,
    redirectUri: redirectUri,
    codeChallenge: codeChallenge,
    codeChallengeMethod: codeChallengeMethod,
    scope: scope,
    expiresAt: DateTime.now().millisecondsSinceEpoch + 60000,
  );
  _authCodes[code] = ac;
  return ac;
}

AuthCode? consumeAuthCode(String code) {
  final ac = _authCodes[code];
  if (ac == null) return null;
  _authCodes.remove(code);
  if (DateTime.now().millisecondsSinceEpoch > ac.expiresAt) return null;
  return ac;
}

AccessToken createAccessToken({
  required String clientId,
  required String scope,
}) {
  final token = _generateRandomString(32);
  final refreshToken = _generateRandomString(32);

  final at = AccessToken(
    token: token,
    clientId: clientId,
    scope: scope,
    expiresAt: DateTime.now().millisecondsSinceEpoch + 3600000, // 1 hr
    refreshToken: refreshToken,
  );

  _accessTokens[token] = at;
  _refreshTokens[refreshToken] = RefreshToken(
    token: refreshToken,
    clientId: clientId,
    scope: scope,
  );

  return at;
}

AccessToken? validateAccessToken(String token) {
  final at = _accessTokens[token];
  if (at == null) return null;
  if (DateTime.now().millisecondsSinceEpoch > at.expiresAt) {
    _accessTokens.remove(token);
    return null;
  }
  return at;
}

RefreshToken? consumeRefreshToken(String token) {
  final rt = _refreshTokens[token];
  if (rt == null) return null;
  _refreshTokens.remove(token);
  return rt;
}

bool revokeToken(String token) {
  if (_accessTokens.remove(token) != null) return true;
  if (_refreshTokens.remove(token) != null) return true;
  return false;
}
